File: Traditional-Important-Algorithms/exercises_py/09_rate_limiting/rate05_circuit_breaker.py
# I AM NOT DONE

"""
Exercise: Circuit Breaker Pattern

Circuit Breaker is a fault tolerance pattern that prevents cascading failures
by stopping requests to a failing service and allowing it time to recover.

States:
- Closed: Normal operation, requests pass through
- Open: Service failing, requests rejected immediately (fail fast)
- Half-Open: Testing recovery, limited requests allowed

Your task: Implement a circuit breaker with three states.

Key concepts:
- State transitions based on failure/success thresholds
- Timeout for recovery attempts
- Failure tracking in sliding window
- Fail-fast behavior
"""

import time
from typing import Callable, TypeVar, Generic, List
from enum import Enum


class CircuitState(Enum):
    """Circuit breaker states"""
    CLOSED = "closed"        # Normal operation
    OPEN = "open"           # Failing, rejecting requests
    HALF_OPEN = "half_open"  # Testing recovery


T = TypeVar('T')
E = TypeVar('E')


class RequestResult:
    """Record of a request result"""

    def __init__(self, success: bool, timestamp: float):
        self.success = success
        self.timestamp = timestamp


class CircuitBreaker:
    """Circuit breaker for fault tolerance"""

    def __init__(self,
                 failure_threshold: int,
                 success_threshold: int,
                 timeout: float,
                 window_size: int):
        """
        Initialize circuit breaker

        Args:
            failure_threshold: Failures needed to open circuit
            success_threshold: Successes needed to close from half-open
            timeout: Time to wait before half-open (in seconds)
            window_size: Number of recent requests to track
        """
        # TODO: Initialize circuit breaker in CLOSED state
        pass

    def _count_recent_failures(self) -> int:
        """Count failures in the recent window"""
        # TODO: Count failures in the recent window
        # - Iterate through recent_results
        # - Count how many have success = False
        # - Only consider up to window_size most recent results
        pass

    def _should_transition_to_half_open(self) -> bool:
        """Check if enough time has passed to try half-open"""
        # TODO: Check if enough time has passed to try half-open
        # - Return True if state is OPEN and timeout has elapsed since opened_at
        # - Otherwise return False
        pass

    def _update_state(self):
        """Update circuit state based on current conditions"""
        # TODO: Update circuit state based on current conditions
        #
        # Logic:
        # 1. If state is OPEN and should_transition_to_half_open:
        #    - Set state to HALF_OPEN
        #    - Reset half_open_successes to 0
        #
        # 2. If state is CLOSED and count_recent_failures >= failure_threshold:
        #    - Set state to OPEN
        #    - Set opened_at to current time
        #
        # 3. If state is HALF_OPEN and half_open_successes >= success_threshold:
        #    - Set state to CLOSED
        #    - Clear recent_results
        pass

    def call(self, operation: Callable[[], T]) -> T:
        """
        Execute operation with circuit breaker protection

        Args:
            operation: Function to execute

        Returns:
            Result of operation

        Raises:
            CircuitBreakerError: If circuit is open
            Exception: If operation raises an exception
        """
        # TODO: Execute operation with circuit breaker protection
        #
        # 1. Update state first
        # 2. If state is OPEN, raise CircuitBreakerError("Circuit breaker is open")
        # 3. Try to execute operation
        # 4. Record result:
        #    - Add to recent_results (keep only last window_size entries)
        #    - If HALF_OPEN and success, increment half_open_successes
        #    - If HALF_OPEN and failure, set state to OPEN with new timeout
        # 5. Update state again
        # 6. Return result or raise exception
        pass

    def state(self) -> CircuitState:
        """Return current state (update state first)"""
        # TODO: Return current state (update state first)
        pass

    def failure_count(self) -> int:
        """Return count of recent failures"""
        # TODO: Return count of recent failures
        pass

    def success_rate(self) -> float:
        """Calculate success rate from recent results"""
        # TODO: Calculate success rate from recent results
        # - If no results, return 1.0 (100%)
        # - Otherwise: successes / total_results
        # - Only consider last window_size results
        pass

    def reset(self):
        """Reset circuit breaker to initial CLOSED state"""
        # TODO: Reset circuit breaker to initial CLOSED state
        # - Clear recent_results
        # - Set state to CLOSED
        # - Reset counters
        pass


class CircuitBreakerError(Exception):
    """Exception raised when circuit breaker is open"""
    pass


import unittest


class TestCircuitBreaker(unittest.TestCase):
    def test_circuit_breaker_creation(self):
        cb = CircuitBreaker(3, 2, 1.0, 10)
        self.assertEqual(cb.state(), CircuitState.CLOSED)

    def test_successful_calls_stay_closed(self):
        cb = CircuitBreaker(3, 2, 1.0, 10)

        for _ in range(10):
            result = cb.call(lambda: 42)
            self.assertEqual(result, 42)

        self.assertEqual(cb.state(), CircuitState.CLOSED)

    def test_failures_open_circuit(self):
        cb = CircuitBreaker(3, 2, 0.1, 10)

        # First 2 failures don't open circuit
        try:
            cb.call(lambda: (_ for _ in ()).throw(Exception("error")))
        except:
            pass
        try:
            cb.call(lambda: (_ for _ in ()).throw(Exception("error")))
        except:
            pass
        self.assertEqual(cb.state(), CircuitState.CLOSED)

        # Third failure opens circuit
        try:
            cb.call(lambda: (_ for _ in ()).throw(Exception("error")))
        except:
            pass
        self.assertEqual(cb.state(), CircuitState.OPEN)

    def test_open_circuit_rejects_requests(self):
        cb = CircuitBreaker(2, 2, 1.0, 10)

        # Open the circuit
        for _ in range(2):
            try:
                cb.call(lambda: (_ for _ in ()).throw(Exception("error")))
            except:
                pass

        self.assertEqual(cb.state(), CircuitState.OPEN)

        # Next request should be rejected without calling operation
        with self.assertRaises(CircuitBreakerError):
            cb.call(lambda: 42)

    def test_transition_to_half_open(self):
        cb = CircuitBreaker(2, 2, 0.1, 10)

        # Open circuit
        for _ in range(2):
            try:
                cb.call(lambda: (_ for _ in ()).throw(Exception("error")))
            except:
                pass

        self.assertEqual(cb.state(), CircuitState.OPEN)

        # Wait for timeout
        time.sleep(0.15)

        # Should transition to half-open on next state check
        self.assertEqual(cb.state(), CircuitState.HALF_OPEN)

    def test_half_open_success_closes_circuit(self):
        cb = CircuitBreaker(2, 2, 0.1, 10)

        # Open circuit
        for _ in range(2):
            try:
                cb.call(lambda: (_ for _ in ()).throw(Exception("error")))
            except:
                pass

        # Wait and transition to half-open
        time.sleep(0.15)
        cb.state()

        self.assertEqual(cb.state(), CircuitState.HALF_OPEN)

        # Successful calls should close circuit
        cb.call(lambda: 1)
        cb.call(lambda: 2)

        self.assertEqual(cb.state(), CircuitState.CLOSED)

    def test_half_open_failure_reopens_circuit(self):
        cb = CircuitBreaker(2, 2, 0.1, 10)

        # Open circuit
        for _ in range(2):
            try:
                cb.call(lambda: (_ for _ in ()).throw(Exception("error")))
            except:
                pass

        # Wait and transition to half-open
        time.sleep(0.15)
        cb.state()

        self.assertEqual(cb.state(), CircuitState.HALF_OPEN)

        # Failure in half-open should reopen circuit
        try:
            cb.call(lambda: (_ for _ in ()).throw(Exception("error")))
        except:
            pass

        self.assertEqual(cb.state(), CircuitState.OPEN)

    def test_failure_count(self):
        cb = CircuitBreaker(5, 2, 1.0, 10)

        cb.call(lambda: 1)
        try:
            cb.call(lambda: (_ for _ in ()).throw(Exception("error")))
        except:
            pass
        try:
            cb.call(lambda: (_ for _ in ()).throw(Exception("error")))
        except:
            pass
        cb.call(lambda: 2)

        self.assertEqual(cb.failure_count(), 2)

    def test_success_rate(self):
        cb = CircuitBreaker(5, 2, 1.0, 10)

        cb.call(lambda: 1)
        cb.call(lambda: 2)
        try:
            cb.call(lambda: (_ for _ in ()).throw(Exception("error")))
        except:
            pass
        cb.call(lambda: 3)

        # 3 successes out of 4 = 75%
        self.assertAlmostEqual(cb.success_rate(), 0.75, places=2)

    def test_window_size_limit(self):
        cb = CircuitBreaker(5, 2, 1.0, 3)

        # Add more results than window size
        for _ in range(5):
            cb.call(lambda: 1)

        try:
            cb.call(lambda: (_ for _ in ()).throw(Exception("error")))
        except:
            pass

        # Should only count failure in window of last 3 requests
        self.assertEqual(cb.failure_count(), 1)

    def test_reset(self):
        cb = CircuitBreaker(2, 2, 1.0, 10)

        # Open circuit
        for _ in range(2):
            try:
                cb.call(lambda: (_ for _ in ()).throw(Exception("error")))
            except:
                pass

        self.assertEqual(cb.state(), CircuitState.OPEN)

        # Reset
        cb.reset()

        self.assertEqual(cb.state(), CircuitState.CLOSED)
        self.assertEqual(cb.failure_count(), 0)

    def test_mixed_results(self):
        cb = CircuitBreaker(3, 2, 1.0, 10)

        cb.call(lambda: 1)
        try:
            cb.call(lambda: (_ for _ in ()).throw(Exception("error")))
        except:
            pass
        cb.call(lambda: 2)
        try:
            cb.call(lambda: (_ for _ in ()).throw(Exception("error")))
        except:
            pass
        cb.call(lambda: 3)

        # Only 2 failures, should still be closed
        self.assertEqual(cb.state(), CircuitState.CLOSED)
        self.assertEqual(cb.failure_count(), 2)


if __name__ == '__main__':
    unittest.main()

File: Traditional-Important-Algorithms/exercises/09_rate_limiting/rate05_circuit_breaker.rs
// rate05_circuit_breaker.rs
//
// Circuit Breaker is a fault tolerance pattern that prevents cascading failures
// by stopping requests to a failing service and allowing it time to recover.
//
// States:
// - Closed: Normal operation, requests pass through
// - Open: Service failing, requests rejected immediately (fail fast)
// - Half-Open: Testing recovery, limited requests allowed
//
// Your task: Implement a circuit breaker with three states.
//
// Key concepts:
// - State transitions based on failure/success thresholds
// - Timeout for recovery attempts
// - Failure tracking in sliding window
// - Fail-fast behavior

// I AM NOT DONE

use std::time::{Duration, Instant};

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum CircuitState {
    Closed,    // Normal operation
    Open,      // Failing, rejecting requests
    HalfOpen,  // Testing recovery
}

#[derive(Debug, Clone, Copy)]
struct RequestResult {
    success: bool,
    timestamp: Instant,
}

pub struct CircuitBreaker {
    state: CircuitState,
    failure_threshold: usize,    // Failures needed to open circuit
    success_threshold: usize,    // Successes needed to close from half-open
    timeout: Duration,           // Time to wait before half-open
    window_size: usize,          // Number of recent requests to track

    recent_results: Vec<RequestResult>,
    opened_at: Option<Instant>,
    half_open_successes: usize,
}

impl CircuitBreaker {
    pub fn new(
        failure_threshold: usize,
        success_threshold: usize,
        timeout: Duration,
        window_size: usize,
    ) -> Self {
        // TODO: Initialize circuit breaker in Closed state
        todo!()
    }

    fn count_recent_failures(&self) -> usize {
        // TODO: Count failures in the recent window
        // - Iterate through recent_results
        // - Count how many have success = false
        // - Only consider up to window_size most recent results
        todo!()
    }

    fn should_transition_to_half_open(&self) -> bool {
        // TODO: Check if enough time has passed to try half-open
        // - Return true if state is Open and timeout has elapsed since opened_at
        // - Otherwise return false
        todo!()
    }

    fn update_state(&mut self) {
        // TODO: Update circuit state based on current conditions
        //
        // Logic:
        // 1. If state is Open and should_transition_to_half_open:
        //    - Set state to HalfOpen
        //    - Reset half_open_successes to 0
        //
        // 2. If state is Closed and count_recent_failures >= failure_threshold:
        //    - Set state to Open
        //    - Set opened_at to current time
        //
        // 3. If state is HalfOpen and half_open_successes >= success_threshold:
        //    - Set state to Closed
        //    - Clear recent_results
        todo!()
    }

    pub fn call<F, T, E>(&mut self, operation: F) -> Result<T, CircuitBreakerError<E>>
    where
        F: FnOnce() -> Result<T, E>,
    {
        // TODO: Execute operation with circuit breaker protection
        //
        // 1. Update state first
        // 2. If state is Open, return Err(CircuitBreakerError::Open)
        // 3. Execute operation
        // 4. Record result:
        //    - Add to recent_results (keep only last window_size entries)
        //    - If HalfOpen and success, increment half_open_successes
        //    - If HalfOpen and failure, set state to Open with new timeout
        // 5. Update state again
        // 6. Return result (wrapped in CircuitBreakerError::Inner if error)
        todo!()
    }

    pub fn state(&mut self) -> CircuitState {
        // TODO: Return current state (update state first)
        todo!()
    }

    pub fn failure_count(&self) -> usize {
        // TODO: Return count of recent failures
        todo!()
    }

    pub fn success_rate(&self) -> f64 {
        // TODO: Calculate success rate from recent results
        // - If no results, return 1.0 (100%)
        // - Otherwise: successes / total_results
        // - Only consider last window_size results
        todo!()
    }

    pub fn reset(&mut self) {
        // TODO: Reset circuit breaker to initial Closed state
        // - Clear recent_results
        // - Set state to Closed
        // - Reset counters
        todo!()
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum CircuitBreakerError<E> {
    Open,       // Circuit is open, request rejected
    Inner(E),   // Underlying operation failed
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_circuit_breaker_creation() {
        let mut cb = CircuitBreaker::new(3, 2, Duration::from_secs(1), 10);
        assert_eq!(cb.state(), CircuitState::Closed);
    }

    #[test]
    fn test_successful_calls_stay_closed() {
        let mut cb = CircuitBreaker::new(3, 2, Duration::from_secs(1), 10);

        for _ in 0..10 {
            let result = cb.call(|| Ok::<_, ()>(42));
            assert_eq!(result, Ok(42));
        }

        assert_eq!(cb.state(), CircuitState::Closed);
    }

    #[test]
    fn test_failures_open_circuit() {
        let mut cb = CircuitBreaker::new(3, 2, Duration::from_millis(100), 10);

        // First 2 failures don't open circuit
        cb.call(|| Err::<(), _>("error")).ok();
        cb.call(|| Err::<(), _>("error")).ok();
        assert_eq!(cb.state(), CircuitState::Closed);

        // Third failure opens circuit
        cb.call(|| Err::<(), _>("error")).ok();
        assert_eq!(cb.state(), CircuitState::Open);
    }

    #[test]
    fn test_open_circuit_rejects_requests() {
        let mut cb = CircuitBreaker::new(2, 2, Duration::from_secs(1), 10);

        // Open the circuit
        cb.call(|| Err::<(), _>("error")).ok();
        cb.call(|| Err::<(), _>("error")).ok();
        assert_eq!(cb.state(), CircuitState::Open);

        // Next request should be rejected without calling operation
        let result = cb.call(|| Ok::<_, ()>(42));
        assert_eq!(result, Err(CircuitBreakerError::Open));
    }

    #[test]
    fn test_transition_to_half_open() {
        let mut cb = CircuitBreaker::new(2, 2, Duration::from_millis(100), 10);

        // Open circuit
        cb.call(|| Err::<(), _>("error")).ok();
        cb.call(|| Err::<(), _>("error")).ok();
        assert_eq!(cb.state(), CircuitState::Open);

        // Wait for timeout
        std::thread::sleep(Duration::from_millis(150));

        // Should transition to half-open on next state check
        assert_eq!(cb.state(), CircuitState::HalfOpen);
    }

    #[test]
    fn test_half_open_success_closes_circuit() {
        let mut cb = CircuitBreaker::new(2, 2, Duration::from_millis(100), 10);

        // Open circuit
        cb.call(|| Err::<(), _>("error")).ok();
        cb.call(|| Err::<(), _>("error")).ok();

        // Wait and transition to half-open
        std::thread::sleep(Duration::from_millis(150));
        cb.state();

        assert_eq!(cb.state(), CircuitState::HalfOpen);

        // Successful calls should close circuit
        cb.call(|| Ok::<_, ()>(1)).ok();
        cb.call(|| Ok::<_, ()>(2)).ok();

        assert_eq!(cb.state(), CircuitState::Closed);
    }

    #[test]
    fn test_half_open_failure_reopens_circuit() {
        let mut cb = CircuitBreaker::new(2, 2, Duration::from_millis(100), 10);

        // Open circuit
        cb.call(|| Err::<(), _>("error")).ok();
        cb.call(|| Err::<(), _>("error")).ok();

        // Wait and transition to half-open
        std::thread::sleep(Duration::from_millis(150));
        cb.state();

        assert_eq!(cb.state(), CircuitState::HalfOpen);

        // Failure in half-open should reopen circuit
        cb.call(|| Err::<(), _>("error")).ok();

        assert_eq!(cb.state(), CircuitState::Open);
    }

    #[test]
    fn test_failure_count() {
        let mut cb = CircuitBreaker::new(5, 2, Duration::from_secs(1), 10);

        cb.call(|| Ok::<_, ()>(1)).ok();
        cb.call(|| Err::<(), _>("error")).ok();
        cb.call(|| Err::<(), _>("error")).ok();
        cb.call(|| Ok::<_, ()>(2)).ok();

        assert_eq!(cb.failure_count(), 2);
    }

    #[test]
    fn test_success_rate() {
        let mut cb = CircuitBreaker::new(5, 2, Duration::from_secs(1), 10);

        cb.call(|| Ok::<_, ()>(1)).ok();
        cb.call(|| Ok::<_, ()>(2)).ok();
        cb.call(|| Err::<(), _>("error")).ok();
        cb.call(|| Ok::<_, ()>(3)).ok();

        // 3 successes out of 4 = 75%
        assert!((cb.success_rate() - 0.75).abs() < 0.01);
    }

    #[test]
    fn test_window_size_limit() {
        let mut cb = CircuitBreaker::new(5, 2, Duration::from_secs(1), 3);

        // Add more results than window size
        for _ in 0..5 {
            cb.call(|| Ok::<_, ()>(1)).ok();
        }

        cb.call(|| Err::<(), _>("error")).ok();

        // Should only count failure in window of last 3 requests
        assert_eq!(cb.failure_count(), 1);
    }

    #[test]
    fn test_reset() {
        let mut cb = CircuitBreaker::new(2, 2, Duration::from_secs(1), 10);

        // Open circuit
        cb.call(|| Err::<(), _>("error")).ok();
        cb.call(|| Err::<(), _>("error")).ok();
        assert_eq!(cb.state(), CircuitState::Open);

        // Reset
        cb.reset();

        assert_eq!(cb.state(), CircuitState::Closed);
        assert_eq!(cb.failure_count(), 0);
    }

    #[test]
    fn test_mixed_results() {
        let mut cb = CircuitBreaker::new(3, 2, Duration::from_secs(1), 10);

        cb.call(|| Ok::<_, ()>(1)).ok();
        cb.call(|| Err::<(), _>("error")).ok();
        cb.call(|| Ok::<_, ()>(2)).ok();
        cb.call(|| Err::<(), _>("error")).ok();
        cb.call(|| Ok::<_, ()>(3)).ok();

        // Only 2 failures, should still be closed
        assert_eq!(cb.state(), CircuitState::Closed);
        assert_eq!(cb.failure_count(), 2);
    }

    #[test]
    fn test_error_propagation() {
        let mut cb = CircuitBreaker::new(5, 2, Duration::from_secs(1), 10);

        let result = cb.call(|| Err::<(), _>("custom error"));
        assert_eq!(result, Err(CircuitBreakerError::Inner("custom error")));
    }
}
